ad9959/src/lib.rs

```diff
@@ -522,14 +479,143 @@ where
+    pub fn freeze(self) -> (I, DdsConfig) {
+        let config = DdsConfig {
+            mode: self.communication_mode,
+        };
+        (self.interface, config)
+    }
+}
+
+/// The frozen DDS configuration.
+pub struct DdsConfig {
+    mode: Mode,
+}
+
+impl DdsConfig {
+    /// Create a serializer that can be used for generating a serialized DDS profile for writing to
+    /// a QSPI stream.
+    pub fn builder(&self) -> ProfileSerializer {
+        ProfileSerializer::new(self.mode)
+    }
+}
+
+/// Represents a means of serializing a DDS profile for writing to a stream.
+pub struct ProfileSerializer {
+    data: [u8; 16],
+    index: usize,
+    mode: Mode,
+}
+
+impl ProfileSerializer {
+    /// Construct a new serializer.
+    ///
+    /// # Args
+    /// * `mode` - The communication mode of the DDS.
+    fn new(mode: Mode) -> Self {
+        Self {
+            mode,
+            data: [0; 16],
+            index: 0,
+        }
+    }
+
+    /// Update a number of channels with the requested profile.
+    ///
+    /// # Args
+    /// * `channels` - A list of channels to apply the configuration to.
+    /// * `ftw` - If provided, indicates a frequency tuning word for the channels.
+    /// * `pow` - If provided, indicates a phase offset word for the channels.
+    /// * `acr` - If provided, indicates the amplitude control register for the channels.
+    pub fn update_channels(
+        &mut self,
+        channels: &[Channel],
+        ftw: Option<u32>,
+        pow: Option<u16>,
+        acr: Option<u16>,
+    ) {
+        // The user should have provided something to update.
+        assert!(
+            (ftw.is_some() || acr.is_some() || pow.is_some())
+                && channels.len() > 0
+        );
+
+        let mut csr: u8 = *0u8.set_bits(1..3, self.mode as u8);
+        for channel in channels.iter() {
+            csr.set_bit(4 + *channel as usize, true);
+        }
+
+        self.add_write(Register::CSR, &[csr]);
+
+        if let Some(ftw) = ftw {
+            self.add_write(Register::CFTW0, &ftw.to_be_bytes());
+        }
+
+        if let Some(pow) = pow {
+            self.add_write(Register::CPOW0, &pow.to_be_bytes());
+        }
+
+        if let Some(acr) = acr {
+            self.add_write(Register::ACR, &acr.to_be_bytes());
+        }
+    }
+
+    /// Add a register write to the serialization data.
+    fn add_write(&mut self, register: Register, value: &[u8]) {
+        let data = &mut self.data[self.index..];
+        assert!(value.len() + 1 <= data.len());
+
+        data[0] = register as u8;
+        data[1..][..value.len()].copy_from_slice(value);
+        self.index += value.len() + 1;
+    }
+
+    /// Get the serialized profile as a slice of 32-bit words.
+    ///
+    /// # Note
+    /// The serialized profile will be padded to the next 32-bit word boundary by adding dummy
+    /// writes to the CSR or FR2 registers.
```

Comment:
I think `LSRR` is better for the 3-byte padding. I don't that ramping being used in streaming mode but `FR2 != 0` might be useful in the future in streaming mode.

Reply:
I agree FR2 isn't the best register - I've updated to LSRR, thanks for the suggestion